numpy/core/src/multiarray/compiled_base.c

```diff
@@ -1103,23 +1103,37 @@ arr_unravel_index(PyObject *self, PyObject *args, PyObject *kwds)
 
     unravel_size = PyArray_MultiplyList(dimensions.ptr, dimensions.len);
 
+    dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        goto fail;
+    }
+
     if (!PyArray_Check(indices0)) {
         indices = (PyArrayObject*)PyArray_FromAny(indices0,
                                                     NULL, 0, 0, 0, NULL);
         if (indices == NULL) {
             goto fail;
         }
+
+        if (PyArray_SIZE(indices) == 0) {
```

Comment:
My personal tendency is to keep this. Seeing it more as a convenience for the user by using something like "`np.asarray(..., when_in_doubt_use=np.intp)`". Not sure if you could actual manage to write it like that, hmmmm.